File: backend/core/action_status.py
"""Canonical Action-Item status vocabulary shared across modules.

One source of truth for the 8 statuses used by Action Center, CTT, Lifestyle
and the Action-Plan steps of Pros & Cons / My Dezider / Solution Finder.
Legacy values ('open', 'in_progress', 'completed') are transparently aliased so
older documents keep rendering correctly.
"""

CANONICAL_STATUSES = [
    "pending", "wip_25", "wip_50", "wip_75",
    "done", "deferred", "blocked", "cancelled",
]

# Only the progress-bearing states carry a numeric % (the rest are terminal/other).
STATUS_PROGRESS = {
    "pending": 0, "wip_25": 25, "wip_50": 50, "wip_75": 75, "done": 100,
}

STATUS_LABELS = {
    "pending": "Pending",
    "wip_25": "WIP 25%",
    "wip_50": "WIP 50%",
    "wip_75": "WIP 75%",
    "done": "Done",
    "deferred": "Deferred",
    "blocked": "Blocked",
    "cancelled": "Cancelled",
}

_ALIASES = {
    "open": "pending",
    "in_progress": "wip_50",
    "in progress": "wip_50",
    "completed": "done",
    "complete": "done",
    "": "pending",
}

_VALID = set(CANONICAL_STATUSES)


def normalize_status(s) -> str:
    """Coerce any incoming status string to a canonical value (default pending)."""
    if s is None:
        return "pending"
    v = str(s).strip().lower()
    if v in _VALID:
        return v
    return _ALIASES.get(v, "pending")


def status_label(s) -> str:
    return STATUS_LABELS.get(normalize_status(s), "Pending")


def progress_for(s) -> int:
    return STATUS_PROGRESS.get(normalize_status(s), 0)

File: frontend/src/constants/actionStatus.ts
/**
 * Canonical Action-Item status vocabulary — one source of truth shared by
 * Action Center, CTT, Lifestyle and the Action-Plan steps of Pros & Cons /
 * My Dezider / Solution Finder. Legacy values are aliased so old data renders.
 */

export type ActionStatus =
  | 'pending' | 'wip_25' | 'wip_50' | 'wip_75'
  | 'done' | 'deferred' | 'blocked' | 'cancelled';

export type StatusOpt = { id: ActionStatus; label: string; color: string; progress?: number };

export const ACTION_STATUS_OPTS: StatusOpt[] = [
  { id: 'pending',   label: 'Pending',   color: '#94A3B8', progress: 0 },
  { id: 'wip_25',    label: 'WIP 25%',   color: '#60A5FA', progress: 25 },
  { id: 'wip_50',    label: 'WIP 50%',   color: '#3B82F6', progress: 50 },
  { id: 'wip_75',    label: 'WIP 75%',   color: '#6366F1', progress: 75 },
  { id: 'done',      label: 'Done',      color: '#10B981', progress: 100 },
  { id: 'deferred',  label: 'Deferred',  color: '#A855F7' },
  { id: 'blocked',   label: 'Blocked',   color: '#EF4444' },
  { id: 'cancelled', label: 'Cancelled', color: '#A1A1AA' },
];

const BY_ID: Record<string, StatusOpt> = ACTION_STATUS_OPTS.reduce(
  (m, o) => { m[o.id] = o; return m; }, {} as Record<string, StatusOpt>);

const ALIAS: Record<string, ActionStatus> = {
  open: 'pending',
  in_progress: 'wip_50',
  'in progress': 'wip_50',
  completed: 'done',
  complete: 'done',
  '': 'pending',
};

export function normStatus(s?: string | null): ActionStatus {
  if (!s) return 'pending';
  const v = String(s).trim().toLowerCase();
  if (BY_ID[v]) return v as ActionStatus;
  return ALIAS[v] || 'pending';
}

export function statusLabel(s?: string | null): string {
  return BY_ID[normStatus(s)].label;
}

export function statusColor(s?: string | null): string {
  return BY_ID[normStatus(s)].color;
}

export function statusProgress(s?: string | null): number {
  return BY_ID[normStatus(s)].progress ?? 0;
}
